### PR TITLE
AMS iAQ Core CO2 sensor component

### PR DESCRIPTION
**Pull request by yozik04**
# What does this implement/fix?
Adds support for AMS iAQ Core CO2 sensors.

Datasheet: https://www.sciosense.com/wp-content/uploads/documents/iaQ-Core-Datasheet.pdf

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3104

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
sensor:
  - platform: iaqcore
    co2:
      name: "iAQ Core CO2 Sensor"
    tvoc:
      name: "iAQ Core TVOC Sensor"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/iaqcore/sensor.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+
+    if CONF_CO2 in config:
+        sens = await sensor.new_sensor(config[CONF_CO2])
+        cg.add(var.set_co2(sens))
+
+    if CONF_TVOC in config:
+        sens = await sensor.new_sensor(config[CONF_TVOC])
+        cg.add(var.set_tvoc(sens))
```
Comment (jesserockz) [inline esphome/components/iaqcore/sensor.py]:
```suggestion
    if co2_config := config.get(CONF_CO2):
        sens = await sensor.new_sensor(co2_config)
        cg.add(var.set_co2(sens))

    if tvoc_config := config.get(CONF_TVOC):
        sens = await sensor.new_sensor(tvoc_config)
        cg.add(var.set_tvoc(sens))
```

File: esphome/components/iaqcore/iaqcore.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+struct SensorData {
+  uint16_t co2;
+  IAQCoreErrorCode status;
+  int32_t resistance;
+  uint16_t tvoc;
+
+  SensorData(const uint8_t *buffer) {
+    this->co2 = (buffer[0] << 8) | buffer[1];
+    this->status = static_cast<IAQCoreErrorCode>(buffer[2]);
+    this->resistance = (buffer[3] << 24) | (buffer[4] << 16) | (buffer[5] << 8) | buffer[6];
```
Comment (jesserockz) [inline esphome/components/iaqcore/iaqcore.cpp]:
```suggestion
    this->resistance = encode_uint32_t(buffer[3], buffer[4], buffer[5], buffer[6]);
```

**Review by jesserockz — DISMISSED** [review]
Mostly looks good, a few code style things =)

**Comment by yozik04** [comment]
@jesserockz I do not see that encode_uint32_t is defined anywhere... Compilation fails as well. I have not found any integration that uses that function.

**Comment by yozik04** [comment]
@jesserockz, I have reverted one of your proposed changes. Can you please approve?

File: esphome/components/iaqcore/iaqcore.cpp (hunk 2 above)
Comment (jesserockz) [reply esphome/components/iaqcore/iaqcore.cpp]: Woops, this is meant to be this, also with a new include for `helpers.h`
```suggestion
    this->resistance = encode_uint32(buffer[3], buffer[4], buffer[5], buffer[6]);
```

**Review by jesserockz — DISMISSED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**